runtime/parachains/src/disputes.rs

```diff
@@ -59,49 +64,60 @@ pub enum DisputeResult {
 	Invalid,
 }
 
-/// Reward hooks for disputes.
-pub trait RewardValidators {
-	// Give each validator a reward, likely small, for participating in the dispute.
-	fn reward_dispute_statement(
-		session: SessionIndex,
-		validators: impl IntoIterator<Item = ValidatorIndex>,
-	);
-}
-
-impl RewardValidators for () {
-	fn reward_dispute_statement(_: SessionIndex, _: impl IntoIterator<Item = ValidatorIndex>) {}
-}
-
 /// Punishment hooks for disputes.
-pub trait PunishValidators {
+pub trait SlashingHandler<BlockNumber> {
 	/// Punish a series of validators who were for an invalid parablock. This is expected to be a major
 	/// punishment.
 	fn punish_for_invalid(
 		session: SessionIndex,
-		validators: impl IntoIterator<Item = ValidatorIndex>,
+		candidate_hash: CandidateHash,
+		losers: impl IntoIterator<Item = ValidatorIndex>,
+		winners: impl IntoIterator<Item = ValidatorIndex>,
```

Comment:
Because currently punishing losers and rewarding the winners is combined due to some limitations in staking. I'll document that.

Reply:
While adding docs for this:
> /// Currently, it's not possible to use
/// `<pallet_staking::Pallet<T>>::reward_by_ids` to reward validators in a
/// previous session (era) if they are no longer in the active set.
/// For this reason, we're currently combining slashing and rewarding in one
/// interface. The rewards are distributed from (a fraction of) the slashing
/// pot.

I couldn't help but think we should just use `reward_by_ids` if the validators are also in the active set. That means in case of disputes, if validators who voted on the right side of the dispute are also in the current active set, they will be rewarded. If they are not in the current active set, they won't (as opposed to the current implementation). This would somewhat simplify the logic and doesn't seem to be gameable.

We can set the `reporters` param when reporting an offence to the current block author as with BABE and GRANDPA equivocations.

@kianenigma any thoughts?